src/main/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtension.java

```diff
@@ -95,93 +126,87 @@ String getParcelableCode(Context context, String className, String classToExtend
-        env.getElementUtils(), Serializable.class);
-    for (ExecutableElement name : properties.values()) {
-      TypeMirror type = name.getReturnType();
-      if (type.getKind() == TypeKind.DECLARED && typeUtils.isSubtype(type, list)) {
-        DeclaredType dType = (DeclaredType) type;
-        List<? extends TypeMirror> types = dType.getTypeArguments();
-        if (!types.isEmpty()) {
-          type = types.get(0);
-        }
-      } else if (type.getKind() == TypeKind.ARRAY) {
+    for (Property property : properties) {
+      TypeMirror type = property.element.getReturnType();
+      if (type.getKind() == TypeKind.ARRAY) {
         ArrayType aType = (ArrayType) type;
         type = aType.getComponentType();
       }
-
-      if (!TypeName.get(type).isPrimitive() &&
-          !TypeSimplifier.isClassOfType(env.getTypeUtils(), parcelable, type) &&
-          !TypeSimplifier.isClassOfType(env.getTypeUtils(), serializable, type)) {
+      TypeElement element = (TypeElement) typeUtils.asElement(type);
+      if ((element == null || !Parcelables.isValidType(typeUtils, element)) &&
+          !Parcelables.isValidType(TypeName.get(type))){
         env.getMessager().printMessage(Diagnostic.Kind.ERROR, "AutoValue property " +
-            name.getSimpleName() + " is not primitive, Parcelable, or Serializable.", name);
+            property.name + " is not a supported Parcelable type.", property.element);
         throw new AutoValueParcelException();
       }
     }
   }
 
+  private static final List<String> METHOD_EXCLUDES =
+      Arrays.asList("describeContents", "toBuilder");
+
   private Map<String, TypeName> convertPropertiesToTypes(
       Map<String, ExecutableElement> properties) {
     Map<String, TypeName> types = new LinkedHashMap<String, TypeName>();
     for (Map.Entry<String, ExecutableElement> entry : properties.entrySet()) {
-      if ("describeContents".equals(entry.getKey())) continue;
+      if (METHOD_EXCLUDES.contains(entry.getKey())) continue;
       types.put(entry.getKey(), TypeName.get(entry.getValue().getReturnType()));
     }
     return types;
   }
 
-  MethodSpec generateConstructor(Map<String, TypeName> properties) {
-    List<ParameterSpec> params = Lists.newArrayList();
-    for (Map.Entry<String, TypeName> entry : properties.entrySet()) {
-      params.add(ParameterSpec.builder(entry.getValue(), entry.getKey()).build());
+  MethodSpec generateConstructor(List<Property> properties) {
+    List<ParameterSpec> params = Lists.newArrayListWithCapacity(properties.size());
+    for (Property property : properties) {
+      params.add(ParameterSpec.builder(property.type, property.name).build());
     }
 
     MethodSpec.Builder builder = MethodSpec.constructorBuilder()
         .addParameters(params);
 
     StringBuilder superFormat = new StringBuilder("super(");
-    for (int i = properties.size(); i > 0; i--) {
+    List<ParameterSpec> args = new LinkedList<ParameterSpec>();
```

Comment:
Who knows.  I sometimes use `LInkedList` over `ArrayList` if I'm only using the collection for iteration, but that doesn't seem to apply here.  Code might have just evolved that way.

In other news, my understanding is that if you're just using iteration and don't need direct access then `LinkedList` is more performant.  Am I wrong in that regard?  I know they offer O(1) performance for iterator adds and removes.